From fe11731407fb5809211ae904fd7cbe13a3103108 Mon Sep 17 00:00:00 2001
From: Bhagavathi Perumal S <bperumal@codeaurora.org>
Date: Fri, 13 Oct 2017 19:45:42 +0530
Subject: [PATCH] ath10k: Add wmi interface to send qos null func

This add wmi interface support to send QOS null func and report the instant
rssi from wmi event WMI_INST_RSSI_STATS_EVENTID to user.

Signed-off-by: Bhagavathi Perumal S <bperumal@codeaurora.org>
---
 drivers/net/wireless/ath/ath10k/debug.c   |    3 +-
 drivers/net/wireless/ath/ath10k/mac.c     |   25 +++++++++++++++
 drivers/net/wireless/ath/ath10k/wmi-ops.h |    7 +++--
 drivers/net/wireless/ath/ath10k/wmi-tlv.c |    3 +-
 drivers/net/wireless/ath/ath10k/wmi.c     |   49 +++++++++++++++++++++++++++--
 drivers/net/wireless/ath/ath10k/wmi.h     |    6 ++++
 6 files changed, 85 insertions(+), 8 deletions(-)

--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -451,7 +451,8 @@ static int ath10k_debug_fw_stats_request
 
 		reinit_completion(&ar->debug.fw_stats_complete);
 
-		ret = ath10k_wmi_request_stats(ar, ar->fw_stats_req_mask);
+		ret = ath10k_wmi_request_stats(ar, ar->fw_stats_req_mask,
+					       (u32)-1, NULL);
 		if (ret) {
 			ath10k_warn(ar, "could not request stats (%d)\n", ret);
 			return ret;
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -7694,6 +7694,30 @@ static void ath10k_mac_op_sta_pre_rcu_re
 			peer->removed = true;
 }
 
+static int
+ath10k_send_qos_null_func(struct ieee80211_hw *hw,
+			  struct ieee80211_vif *vif,
+			  struct ieee80211_sta *sta)
+{
+	struct ath10k *ar = hw->priv;
+	struct ath10k_vif *arvif = (void *)vif->drv_priv;
+	int ret = 0;
+
+	mutex_lock(&ar->conf_mutex);
+
+	ret = ath10k_wmi_request_stats(ar, WMI_10_4_STAT_INST, arvif->vdev_id,
+				       sta->addr);
+	if (ret) {
+		ath10k_warn(ar, "could not request stats (%d)\n", ret);
+		mutex_unlock(&ar->conf_mutex);
+		return ret;
+	}
+
+	mutex_unlock(&ar->conf_mutex);
+	return ret;
+}
+
+
 static const struct ieee80211_ops ath10k_ops = {
 	.tx				= ath10k_mac_op_tx,
 	.wake_tx_queue			= ath10k_mac_op_wake_tx_queue,
@@ -7747,6 +7771,7 @@ static const struct ieee80211_ops ath10k
 	.sta_add_debugfs		= ath10k_sta_add_debugfs,
 	.sta_statistics			= ath10k_sta_statistics,
 #endif
+	.send_qos_null_func		= ath10k_send_qos_null_func,
 };
 
 #define CHAN2G(_channel, _freq, _flags) { \
@@ -8302,6 +8327,7 @@ int ath10k_mac_register(struct ath10k *a
 	ieee80211_hw_set(ar->hw, QUEUE_CONTROL);
 	ieee80211_hw_set(ar->hw, SUPPORTS_TX_FRAG);
 	ieee80211_hw_set(ar->hw, REPORTS_LOW_ACK);
+	ieee80211_hw_set(ar->hw, SEND_NULL_FUNC);
 
 	if (ar->ethernet_mode)
 		ieee80211_hw_set(ar->hw, SUPPORTS_80211_ENCAP);
--- a/drivers/net/wireless/ath/ath10k/wmi-ops.h
+++ b/drivers/net/wireless/ath/ath10k/wmi-ops.h
@@ -120,7 +120,8 @@ struct wmi_ops {
 					  bool deliver_cab);
 	struct sk_buff *(*gen_pdev_set_wmm)(struct ath10k *ar,
 					    const struct wmi_wmm_params_all_arg *arg);
-	struct sk_buff *(*gen_request_stats)(struct ath10k *ar, u32 stats_mask);
+	struct sk_buff *(*gen_request_stats)(struct ath10k *ar, u32 stats_mask,
+					     u32 vdev_id, const u8 *mac);
 	struct sk_buff *(*gen_force_fw_hang)(struct ath10k *ar,
 					     enum wmi_force_fw_hang_type type,
 					     u32 delay_ms);
@@ -940,14 +941,14 @@ ath10k_wmi_pdev_set_wmm_params(struct at
 }
 
 static inline int
-ath10k_wmi_request_stats(struct ath10k *ar, u32 stats_mask)
+ath10k_wmi_request_stats(struct ath10k *ar, u32 stats_mask, u32 vdev_id, u8 *addr)
 {
 	struct sk_buff *skb;
 
 	if (!ar->wmi.ops->gen_request_stats)
 		return -EOPNOTSUPP;
 
-	skb = ar->wmi.ops->gen_request_stats(ar, stats_mask);
+	skb = ar->wmi.ops->gen_request_stats(ar, stats_mask, vdev_id, addr);
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
--- a/drivers/net/wireless/ath/ath10k/wmi-tlv.c
+++ b/drivers/net/wireless/ath/ath10k/wmi-tlv.c
@@ -2431,7 +2431,8 @@ ath10k_wmi_tlv_op_gen_pdev_set_wmm(struc
 }
 
 static struct sk_buff *
-ath10k_wmi_tlv_op_gen_request_stats(struct ath10k *ar, u32 stats_mask)
+ath10k_wmi_tlv_op_gen_request_stats(struct ath10k *ar, u32 stats_mask,
+				    u32 vdev_id, u8 *mac)
 {
 	struct wmi_request_stats_cmd *cmd;
 	struct wmi_tlv *tlv;
--- a/drivers/net/wireless/ath/ath10k/wmi.c
+++ b/drivers/net/wireless/ath/ath10k/wmi.c
@@ -3135,6 +3135,44 @@ void ath10k_wmi_event_update_stats(struc
 	ath10k_debug_fw_stats_process(ar, skb);
 }
 
+void ath10k_wmi_event_update_inst_rssi(struct ath10k *ar, struct sk_buff *skb)
+{
+	const struct wmi_10_4_inst_stats *src;
+	u32 instant_rssi;
+	struct ieee80211_sta *sta;
+	struct ieee80211_tx_info info;
+
+	ath10k_dbg(ar, ATH10K_DBG_WMI, "WMI_INST_RSSI_STATS_EVENTID\n");
+
+	if (skb->len < sizeof(struct wmi_10_4_inst_stats)) {
+		ath10k_warn(ar, "failed to get the inst rssi");
+		return;
+	}
+
+	src = (void *)skb->data;
+	skb_pull(skb,sizeof(struct wmi_10_4_inst_stats));
+
+	instant_rssi = __le32_to_cpu(src->inst_rssi);
+	rcu_read_lock();
+	sta = ieee80211_find_sta_by_ifaddr(ar->hw, src->peer_macaddr.addr,
+						   NULL);
+	if (!sta) {
+		ath10k_warn(ar, "failed invalid sta");
+		goto exit;
+	}
+	if (!instant_rssi){
+		ath10k_warn(ar, "failed invalid inst rssi");
+	} else {
+		memset(&info, 0, sizeof(info));
+		info.flags = IEEE80211_TX_STAT_ACK;
+		info.status.ack_signal = ATH10K_DEFAULT_NOISE_FLOOR +
+					 instant_rssi;
+		ieee80211_tx_status_noskb(ar->hw, sta, &info);
+	}
+exit:
+	rcu_read_unlock();
+}
+
 static int
 ath10k_wmi_op_pull_vdev_start_ev(struct ath10k *ar, struct sk_buff *skb,
 				 struct wmi_vdev_start_ev_arg *arg)
@@ -5523,6 +5561,9 @@ static void ath10k_wmi_10_4_op_rx(struct
 	case WMI_10_4_UPDATE_STATS_EVENTID:
 		ath10k_wmi_event_update_stats(ar, skb);
 		break;
+	case WMI_10_4_INST_RSSI_STATS_EVENTID:
+		ath10k_wmi_event_update_inst_rssi(ar, skb);
+		break;
 	case WMI_10_4_PDEV_TEMPERATURE_EVENTID:
 		ath10k_wmi_event_temperature(ar, skb);
 		break;
@@ -7053,7 +7094,8 @@ ath10k_wmi_op_gen_pdev_set_wmm(struct at
 }
 
 static struct sk_buff *
-ath10k_wmi_op_gen_request_stats(struct ath10k *ar, u32 stats_mask)
+ath10k_wmi_op_gen_request_stats(struct ath10k *ar, u32 stats_mask,
+				u32 vdev_id, const u8 *mac)
 {
 	struct wmi_request_stats_cmd *cmd;
 	struct sk_buff *skb;
@@ -7064,9 +7106,13 @@ ath10k_wmi_op_gen_request_stats(struct a
 
 	cmd = (struct wmi_request_stats_cmd *)skb->data;
 	cmd->stats_id = __cpu_to_le32(stats_mask);
+	if (vdev_id != (u32)-1)
+		cmd->vdev_id = __cpu_to_le32(vdev_id);
+	if (mac)
+		ether_addr_copy(cmd->peer_macaddr.addr, mac);
 
-	ath10k_dbg(ar, ATH10K_DBG_WMI, "wmi request stats 0x%08x\n",
-		   stats_mask);
+	ath10k_dbg(ar, ATH10K_DBG_WMI, "wmi request stats 0x%08x, vdevid=%d, mac=%pM\n",
+		   stats_mask, vdev_id, mac != NULL ? mac : "none");
 	return skb;
 }
 
--- a/drivers/net/wireless/ath/ath10k/wmi.h
+++ b/drivers/net/wireless/ath/ath10k/wmi.h
@@ -4352,6 +4352,7 @@ enum wmi_stats_id {
 	WMI_STAT_VDEV = BIT(3),
 	WMI_STAT_BCNFLT = BIT(4),
 	WMI_STAT_VDEV_RATE = BIT(5),
+	WMI_STAT_INST_RSSI = BIT(6),
 };
 
 enum wmi_10_4_stats_id {
@@ -4596,6 +4597,11 @@ struct wmi_vdev_create_cmd {
 	struct wmi_mac_addr vdev_macaddr;
 } __packed;
 
+struct wmi_10_4_inst_stats {
+	__le32 inst_rssi;
+	struct wmi_mac_addr peer_macaddr;
+} __packed;
+
 enum wmi_vdev_type {
 	WMI_VDEV_TYPE_AP      = 1,
 	WMI_VDEV_TYPE_STA     = 2,
